internal/runner/runner.go

```diff
@@ -0,0 +1,148 @@
+package runner
+
+import (
+	"errors"
+	"os"
+	"path/filepath"
+	"strings"
+	"syscall"
+)
+
+// CommandRunner runs commands and returns the build steps that were run.
+type CommandRunner struct {
+	// Env is global environment variables passed to all commands.
+	Env []string
+
+	// Steps are the steps to execute.
+	Steps []*CommandStep
+}
+
+// CommandStep is a command that was executed by the builder.
+type CommandStep struct {
+	// Command is the command that was run.
+	Command []string `json:"command"`
+
+	// Env are the environment variables passed to the command.
+	Env []string `json:"env"`
+
+	// WorkingDir is the working directory the command was executed in.
+	WorkingDir string `json:"workingDir"`
+}
+
+// Run excutes a series of commands and returns the steps that were executed
+// successfully. Commands are run in sequence and are expected to return a zero
+// exit status.
+//
+// Global environment variables are merged with steps environment variables in
+// the returned steps. In the case of duplicates the last occurrence has precidence.
+// Environment variables are *not* inherited from the current process.
+//
+// The returned CommandSteps should be included in the buildConfig provenance.
+// These are *not* the same as the runner commands. Env vars are sanitized, pwd
+// is changed to the absolute path, and only commands that executed
+// successfully are returned.
+func (r *CommandRunner) Run() (steps []*CommandStep, err error) {
+	var originalwd string
+	originalwd, err = os.Getwd()
+	if err != nil {
+		return // steps, err
+	}
+	defer func() {
+		// Change directories back to the original working directory but only
+		// return the error returned by Chdir if no other error occurred.
+		if chDirErr := os.Chdir(originalwd); err == nil {
+			// NOTE: err is returned by the function after the defer is called.
+			err = chDirErr
+		}
+	}()
+
+	for _, step := range r.Steps {
+		var runStep *CommandStep
+		runStep, err = r.runStep(step)
+		if err != nil {
+			return // steps, err
+		}
+		steps = append(steps, runStep)
+	}
+	return // steps, err
+}
+
+func (r *CommandRunner) runStep(step *CommandStep) (*CommandStep, error) {
+	if len(step.Command) == 0 {
+		return nil, errors.New("command is empty")
+	}
+
+	cmd := step.Command[0]
+	args := step.Command[1:]
+	// TODO: Add some kind of LD_PRELOAD protection?
+	env := dedupEnv(append(r.Env, step.Env...))
+
+	// Set the POSIX PWD env var.
+	pwd, err := filepath.Abs(step.WorkingDir)
+	if err != nil {
+		return nil, err
+	}
+	env = append(env, "PWD="+pwd)
+
+	if err = os.Chdir(pwd); err != nil {
+		return nil, err
+	}
+
+	// NOTE: We use syscall.Exec directly because we are executing the
+	// commands in sequence and we want full control over the environment.
+	// The stdlib exec package deduplicates env vars and so it's hard to
+	// know exactly what the command was run with.
```

Comment:
this command is controlled by the developer, so we cannot trust what it returns, correct? How do we plan to use the returned values?

Reply:
I'm not sure what you're referring to. Do you mean stderr/stdout? Right now I'm thinking that the only thing that would be used would be the `error` value returned. The only real value that impacts that would be the command's exit code. stdout/stderr would just get printed for debugging purposes.

Reply:
I may have added my comment in the wrong place. I meant that:
```
return &CommandStep{
		Command:    append([]string{cmd}, args...),
		Env:        env,
		WorkingDir: pwd,
	}, nil
```

seemed to return the steps to add in the provenance. Since it's never called, I think that was more a test than anything. So you can probably ignore my comment :)

Reply:
So to be clear,

- step output will go to stdout but not parsed by the runner API or builder in any way.
- Only error code is checked to determine if the step was executed successfully or not
- We return separate command steps than those that are given in order to fill in values like working directory and some default environment variables to most accurately describe how the command was executed.